File: src/utils/sanitize.js
// src/utils/sanitize.js
// Small utility helpers for cleaning up AI model responses and basic guards.

/**
 * stripFences(text)
 * Removes triple-backtick fenced code blocks and returns their inner content.
 * If multiple fenced blocks exist, joins them with blank line.
 * If no fences are present, strips stray leading/trailing triple-backticks.
 */
export function stripFences(text = "") {
  if (typeof text !== "string") return "";
  const fenceRegex = /```[a-zA-Z0-9+\-._]*\n([\s\S]*?)```/g;
  const matches = [];
  let m;
  while ((m = fenceRegex.exec(text)) !== null) {
    matches.push(m[1]);
  }
  if (matches.length > 0) {
    return matches.join("\n\n").trim();
  }
  // fallback: remove single leading/trailing fences if present
  return text.replace(/^\s*```[^\n]*\n?/, "").replace(/```\s*$/, "").trim();
}

/**
 * hasDangerousShell(text)
 * Quick heuristic to detect obviously dangerous shell commands in the converted output.
 * This is NOT foolproof — only a lightweight warning for the UI.
 */
export function hasDangerousShell(text = "") {
  if (typeof text !== "string") return false;
  const dangerous = /(?:rm\s+-rf|sudo\b|mkfs\b|dd\s+if=|:(){:|forkbomb|chmod\s+777|curl\s+--|wget\s+--)/i;
  return dangerous.test(text);
}

/**
 * truncateOutput(text, maxLen = 200000)
 * Truncates output for display while returning a truncated notice.
 */
export function truncateOutput(text = "", maxLen = 200000) {
  if (typeof text !== "string") return "";
  if (text.length <= maxLen) return text;
  return text.slice(0, maxLen) + "\n\n// ...truncated (full output available for download) ...";
}

File: devtranspiler/backend/core/sanitize.py
"""
core/sanitize.py
Server-side sanitisation — mirrors the frontend sanitize.js but is the
authoritative check (never trust client-side only).
"""

import re
import hashlib
from typing import Tuple

# ── Constants ──────────────────────────────────────────────────────────────
MAX_INPUT_CHARS = 50_000      # ~1 250 lines of code
MAX_OUTPUT_CHARS = 200_000

# Patterns that suggest the LLM was jailbroken into producing shell commands
_DANGEROUS_SHELL = re.compile(
    r"(?:rm\s+-rf|sudo\b|mkfs\b|dd\s+if=|:\(\)\{:|forkbomb|"
    r"chmod\s+777|curl\s+.*\|\s*(?:sh|bash)|wget\s+.*\|\s*(?:sh|bash))",
    re.IGNORECASE,
)

# Fence patterns the LLM adds around code blocks
_FENCE_RE = re.compile(r"```[a-zA-Z0-9+\-._]*\n([\s\S]*?)```")


# ── Public helpers ─────────────────────────────────────────────────────────

def strip_fences(text: str) -> str:
    """Remove markdown code fences; return inner content."""
    matches = _FENCE_RE.findall(text)
    if matches:
        return "\n\n".join(m.strip() for m in matches)
    # Fallback: strip stray leading/trailing fences
    text = re.sub(r"^\s*```[^\n]*\n?", "", text)
    text = re.sub(r"```\s*$", "", text)
    return text.strip()


def has_dangerous_shell(text: str) -> bool:
    """Heuristic check — informational, not a security guarantee."""
    return bool(_DANGEROUS_SHELL.search(text))


def truncate_output(text: str, max_len: int = MAX_OUTPUT_CHARS) -> str:
    if len(text) <= max_len:
        return text
    return text[:max_len] + "\n\n// ...truncated (output exceeded display limit)..."


def validate_input(code: str, source_lang: str, target_lang: str) -> Tuple[bool, str]:
    """
    Returns (is_valid, error_message).
    Call this before enqueuing a job.
    """
    if not code or not code.strip():
        return False, "Input code cannot be empty."

    if len(code) > MAX_INPUT_CHARS:
        return False, f"Input exceeds {MAX_INPUT_CHARS:,} character limit."

    if source_lang == target_lang:
        return False, "Source and target language must be different."

    return True, ""


def make_cache_key(source_lang: str, target_lang: str, code: str) -> str:
    """
    Deterministic SHA-256 cache key.
    Identical conversions hit cache regardless of whitespace normalisation.
    """
    normalised = " ".join(code.split())   # collapse whitespace
    raw = f"{source_lang}:{target_lang}:{normalised}"
    return "conv:" + hashlib.sha256(raw.encode()).hexdigest()